helix-term/src/config.rs

```diff
@@ -43,11 +35,71 @@ impl Display for ConfigLoadError {
 }
 
 impl Config {
+    // avoids confusion between a default rust struct and default helix keys
+    /// Creates a new empty config
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Config {
+            theme: None,
+            unbind_default_keys: false,
+            keys: HashMap::new(),
+            editor: helix_view::editor::Config::default(),
+        }
+    }
+
+    /// Creates a new config with the default keybindings
+    pub fn with_default_keys() -> Self {
+        Config {
+            theme: None,
+            unbind_default_keys: false,
+            keys: default_keys(),
+            editor: helix_view::editor::Config::default(),
+        }
+    }
+
+    /// Overwrite (or add) other's keys into our keymap
+    pub fn merge_keys(mut self, mut other: Self) -> Self {
+        for (mode, keys) in self.keys.iter_mut() {
+            if let Some(other_keys) = other.keys.remove(mode) {
+                keys.merge(other_keys);
+            }
+        }
+
+        self
+    }
+
+    /// Take other's keys and merge ours into them
+    pub fn merge_into_keys(mut self, mut other: Self) -> Self {
+        std::mem::swap(&mut self.keys, &mut other.keys);
+        self.merge_keys(other)
+    }
+
+    /// Replace our keys for a mode with other's keys for that mode.
+    /// Retains existing modes if other does not redefine them
+    pub fn replace_key_modes(mut self, other: Self) -> Self {
+        for (mode, keys) in other.keys.into_iter() {
+            self.keys.insert(mode, keys);
+        }
+
+        self
+    }
+
+    /// Take other's keys and replace our modes into them
+    pub fn replace_into_key_modes(mut self, mut other: Self) -> Self {
+        std::mem::swap(&mut self.keys, &mut other.keys);
+        self.replace_key_modes(other)
+    }
+
     pub fn load(config_path: PathBuf) -> Result<Config, ConfigLoadError> {
         match std::fs::read_to_string(config_path) {
-            Ok(config) => toml::from_str(&config)
-                .map(merge_keys)
-                .map_err(ConfigLoadError::BadConfig),
+            Ok(config) => match toml::from_str::<Config>(&config) {
+                Ok(config) if config.unbind_default_keys => {
+                    Ok(config.replace_into_key_modes(Config::with_default_keys()))
+                }
+                Ok(config) => Ok(config.merge_into_keys(Config::with_default_keys())),
+
+                Err(err) => Err(ConfigLoadError::BadConfig(err)),
+            },
             Err(err) => Err(ConfigLoadError::Error(err)),
         }
```

Comment:
This can be refactored to reduce the nesting:

```rs
let config_string =
    std::fs::read_to_string(config_path).map_err(|err| ConfigLoadError::Error(err))?;
let config = toml::from_str::<Config>(&config_string)
    .map_err(|err| ConfigLoadError::BadConfig(err))?;

if config.unbind_default_keys {
    Ok(config.replace_into_key_modes(Config::with_default_keys()))
} else {
    Ok(config.merge_into_keys(Config::with_default_keys()))
}
```

Reply:
These methods aren't needed anymore with the current structure of helix